casper/src/main/scala/coop/rchain/casper/Casper.scala

```diff
@@ -68,6 +69,35 @@ trait Casper[F[_]] {
   def getDependencyFreeFromBuffer: F[List[BlockMessage]]
 }
 
+object Casper {
+
+  /**
+    * Casper messages serve only two purposes:
+    * 1. offering state transitions for validation by the network, and
+    * 2. participating in past state transition finalization.
+    * Therefore message producing makes sense and should be allowed only if it is justified by these reasons.
+    *
+    * Given said above here are rules for detecting if propose is justified:
+    * 1. State transition
+    *    a) there is a user deploy to be included in a block, or
+    *    b) parents have different post state, so merge block should be issued.
+    *    Both of these rules can be validated, so offence an be detected.
+    * 2. Finalization (reaching acquiescence)
+    *    There are non finalized messages which do not share post state with some finalized message.
+    *    This rule also can be validated and offence detected.
+    *    TODO this can be narrowed down to "if sender already propagated its weight down to all states known"
+    */
+  def shouldPropose[F[_]: Monad](
+      s: CasperSnapshot[F],
+      deploysWaiting: F[Set[Signed[DeployData]]]
+  ): F[Boolean] = {
+    val hasParentsToMerge = (s.parents.map(_.body.state.postStateHash).distinct.size > 1).pure[F]
+    val hasDeploysToOffer = deploysWaiting.map(_ -- s.deploysInScope).map(_.nonEmpty)
+    val noAcquiescence    = s.dag.reachedAcquiescence.not
+    hasDeploysToOffer ||^ hasParentsToMerge ||^ noAcquiescence
```

Comment:
Do we need to cover slashing deploys also?

Reply:
Yes, good catch!

Reply:
https://github.com/rchain/rchain/pull/3509/commits/914256b244b36f84dd5d12d67d84082612d85ec5